torch/nn/functional.py

```diff
@@ -2582,14 +2582,14 @@ def _pointwise_loss(lambd, lambd_optimized, input, target, reduction='mean'):
         return lambd_optimized(expanded_input, expanded_target, _Reduction.get_enum(reduction))
 
 
-def _smooth_l1_loss(input, target):
-    # type: (Tensor, Tensor) -> Tensor
+def _smooth_l1_loss(input, target, delta=1.):
+    # type: (Tensor, Tensor, float) -> Tensor
     t = torch.abs(input - target)
-    return torch.where(t < 1, 0.5 * t ** 2, t - 0.5)
+    return torch.where(t < delta, 0.5 * t ** 2, t * delta - (0.5 * delta ** 2))
```

Comment:
I believe this should be `torch.where(t <= delta, ..., ...)` (see the wiki article for example).

You could consider this thread for example: [https://github.com/devsisters/DQN-tensorflow/issues/16#issuecomment-306162797](https://github.com/devsisters/DQN-tensorflow/issues/16#issuecomment-306162797)

There, someone argues that loss function for training DQN should be Huber with a delta of 2:
> ... to qualify as correct Q-learning, you want to compute an estimate over possible future states... This what you would get with MSE, or Huber loss with behaving as MSE in [-2, 2]

If the implementation of Huber uses `t < delta`, then that "intuitive" delta value of 2 would still not be large enough for DQN.

Reply:
When `t=delta`, both expressions are the same, so it wouldn't matter which one is picked. Do I understand your point correctly?

In particular, the first is 
```
0.5 * t ** 2
= 0.5 * delta ** 2.
```
The second is 
```
t * delta - (0.5 * delta ** 2)
= delta * delta - (0.5 * delta ** 2)
= 0.5 * delta ** 2,
```
as before.

Reply:
Whoops - good point, you're right. (Another way I should have noticed this - there is no discontinuity in the function, so it must be that the two piecewise components have the same value at the border).

Perhaps it would still be more "obviously correct" to have the formula here match the canonical formula, but you're right that there is not a functional difference. Cheers